curp/src/channel/key_mpsc.rs

```diff
@@ -0,0 +1,359 @@
+                super::keys_conflict(k.keys(), new_km.keys()).then(|| {
+                    let _ignore = v.insert(new_km.clone());
+                    1
+                })
+            })
+            .sum();
+        // the message can only be inserted once, so we ignore the return value
+        let _ignore = self.successor.insert(new_km.clone(), HashSet::new());
+
+        if proceeder_cnt != 0 {
+            // the message can only be inserted once, so we ignore the return value
+            let _ignore2 = self.proceeder.insert(new_km, proceeder_cnt);
+        }
+    }
+
+    /// Append a key and message to this `KeybasedChannel`
+    fn append(&mut self, keys: &[K], msg: M) -> MpscKeysMessage<K, M> {
+        let km = MpscKeysMessage::new(keys.to_vec(), msg);
+        self.insert_graph(km.clone());
+        km
+    }
+
+    /// Move a message for the `key` from pending to inner recursively
+    fn mark_ready(&mut self, km: &MpscKeysMessage<K, M>) {
+        km.mark_ready();
+        let mut ready_pool = VecDeque::new();
+        let mut ready_cnt = 0;
+        if self.proceeder.get(km).is_none() {
+            ready_pool.push_back(km.clone());
+            ready_cnt = 1;
+        }
+
+        while let Some(cur) = ready_pool.pop_front() {
+            self.inner.push_back(cur.clone());
+            if let Some(successor) = self.successor.remove(km) {
+                for s in successor {
+                    let (no_proceeder, has_pre) = if let Some(s_p) = self.proceeder.get_mut(&s) {
+                        *s_p = s_p.overflow_sub(1);
+                        (*s_p == 0, true)
+                    } else {
+                        (true, false)
+                    };
+
+                    if no_proceeder && has_pre {
+                        let _ignore_removed = self.proceeder.remove(&s);
+                    }
+
+                    if no_proceeder && s.ready() {
+                        ready_pool.push_back(s);
+                        ready_cnt = ready_cnt.overflow_add(1);
+                    }
+                }
+            }
+        }
+
+        if ready_cnt > 0 {
+            self.new_msg_event.notify(ready_cnt);
+        }
+    }
+}
+
+/// The sender inner type
+type MpscKeybasedSenderInner<K, M> = KeybasedSenderInner<MpscKeysMessage<K, M>>;
+
+#[derive(Clone)]
+/// The Sender for the `KeybasedChannel`
+pub(crate) struct MpscKeybasedSender<K, M> {
+    /// The channel
+    inner: Arc<MpscKeybasedSenderInner<K, M>>,
+}
+
+impl<K: Eq + Hash + Clone + ConflictCheck + Send + 'static, M: Send + 'static>
+    MpscKeybasedSender<K, M>
+{
+    /// Create a new sender
+    fn new(channel: Arc<Mutex<KeybasedChannel<MpscKeysMessage<K, M>>>>) -> Self {
+        Self {
+            inner: Arc::new(MpscKeybasedSenderInner { channel }),
+        }
+    }
+
+    /// Send a key and message to the channel.
+    /// If the channel closed, returns `None`; otherwise returns a `Notify`
+    #[allow(dead_code)]
```

Comment:
Is this still needed?

Reply:
A good reminder, I need to scan the dead_code again.